src/storage/txn/tracker.rs

```diff
@@ -0,0 +1,302 @@
+// Copyright 2025 TiKV Project Authors. Licensed under Apache-2.0.
+
+use std::cell::RefCell;
+
+use tikv_util::time::Instant;
+use tracker::{FutureTrack, TrackerToken, GLOBAL_TRACKERS};
+
+use crate::storage::metrics::CommandKind;
+
+thread_local! {
+    // The TLS state exists solely for the purpose of [`TlsFutureTracker::collect_to_tracker`]
+    // Remove this complication if we can find a better way to collect the time.
+    static CURRENT_STATE: RefCell<Option<State>> = const { RefCell::new(None) };
+}
+
+/// PollState tracks the state of the future polling.
+#[derive(Debug, Clone, Copy)]
+enum PollState {
+    // The state is set to Collected when [`TlsFutureTracker::collect_to_tracker`]
+    // is called, meaning the process time and suspend time are collected.
+    Collected,
+    // The state is set to Begin when `Future::poll` is called.
+    // Record the time when the future is polled, and the time is used to
+    // calculate the process time of the future.
+    Began(Instant),
+    // The state is set to Finish when `Future::poll` is finished.
+    // Record the time when the future is finished, and the time is used to
+    // calculate the suspend time before the next poll.
+    Finished(Instant),
+}
+
+#[derive(Debug)]
+struct State {
+    token: TrackerToken,
+    tag: CommandKind,
+    cid: u64,
+    current_stage: PollState,
+
+    future_process_nanos: u64,
+    future_suspend_nanos: u64,
+}
+
+pub struct TlsFutureTracker {
+    state: Option<State>,
+}
+
+impl TlsFutureTracker {
+    pub fn new(token: TrackerToken, tag: CommandKind, cid: u64) -> Self {
+        let state = State {
+            token,
+            tag,
+            cid,
+            // Set to Finish state to tracking the wait time before the first poll.
+            current_stage: PollState::Finished(Instant::now()),
+
+            future_process_nanos: 0,
+            future_suspend_nanos: 0,
+        };
+        TlsFutureTracker { state: Some(state) }
+    }
+
+    /// Collects the current future process and suspend time to the tracker.
+    ///
+    /// It exists because a storage callback is invoked during the future
+    /// polling. To make measurements more accurate, we need to collect the
+    /// time before the callback is invoked.
+    pub fn collect_to_tracker(now: Instant, tracker: &mut tracker::Tracker) {
+        CURRENT_STATE.with(|current_state| {
+            let mut state = current_state.borrow_mut();
+            let state = state.as_mut().expect("TLS future tracker state is not set");
+            state.collect_to_tracker(now, tracker);
+            // Prevent double counting.
+            state.current_stage = PollState::Collected;
```

Comment:
NIT: Seems `current_stage` is never read again after set to `Collected`, and there is no code related to this state, do we really need this state?